examples/js/utils/RoughnessMipmapper.js

```diff
@@ -0,0 +1,174 @@
+/**
+ * @author Emmett Lalish / elalish
+ *
+ * This class generates custom mipmaps for a roughness map by encoding the lost variation in the
+ * normal map mip levels as increased roughness in the corresponding roughness mip levels. This
+ * helps with rendering accuracy for MeshPhysicalMaterial, and also helps with anti-aliasing when
+ * using PMREM. If the normal map is larger than the roughness map, the roughness map will be
+ * enlarged to match the dimensions of the normal map.
+ */
+
+THREE.RoughnessMipmapper = ( function () {
+
+	var _mipmapMaterial = _getMipmapMaterial();
+	var _scene = new THREE.Scene();
+	_scene.add( new THREE.Mesh( new THREE.PlaneBufferGeometry( 2, 2 ), _mipmapMaterial ) );
+
+	var _flatCamera = new THREE.OrthographicCamera( 0, 1, 0, 1, 0, 1 );
+	var _tempTarget = null;
+	var _renderer = null;
+
+	// constructor
+	var RoughnessMipmapper = function ( renderer ) {
+
+		_renderer = renderer;
+
+	};
+
+	RoughnessMipmapper.prototype = {
+
+		constructor: RoughnessMipmapper,
+
+		generateMipmaps: function ( material ) {
+
+			var { roughnessMap, normalMap } = material;
+			if ( roughnessMap == null || normalMap == null || ! roughnessMap.generateMipmaps ||
+                material.userData.roughnessUpdated ) return;
+
+			material.userData.roughnessUpdated = true;
+
+			var width = Math.max( roughnessMap.image.width, normalMap.image.width );
+			var height = Math.max( roughnessMap.image.height, normalMap.image.height );
+			if ( ! THREE.Math.isPowerOfTwo( width ) || ! THREE.Math.isPowerOfTwo( height ) ) return;
+
+			var autoClear = _renderer.autoClear;
+			_renderer.autoClear = false;
+
+			if ( _tempTarget == null || _tempTarget.width !== width || _tempTarget.height !== height ) {
+
+				if ( _tempTarget != null ) _tempTarget.dispose();
+
+				_tempTarget = new THREE.WebGLRenderTarget( width, height, { depthBuffer: false, stencilBuffer: false } );
+
+			}
+
+			if ( width !== roughnessMap.image.width || height !== roughnessMap.image.height ) {
+
+				var newRoughnessTarget = new THREE.WebGLRenderTarget( width, height, {
+					minFilter: THREE.LinearMipMapLinearFilter,
+					depthBuffer: false,
+					stencilBuffer: false
+				} );
+				newRoughnessTarget.texture.generateMipmaps = true;
+				// Setting the render target causes the memory to be allocated.
+				_renderer.setRenderTarget( newRoughnessTarget );
+				material.roughnessMap = newRoughnessTarget.texture;
+				if ( material.metalnessMap != null ) material.metalnessMap = material.roughnessMap;
+				if ( material.aoMap != null ) material.aoMap = material.roughnessMap;
```

Comment:
While glTF guarantees that roughness maps will always be packed with the metalness map (if any), three.js does not. And the AO map could be a separate texture in either glTF or three.js. So I think this check should verify that the original metalness and AO maps were the same texture as the roughness map before reassigning.

Reply:
Thanks, I didn't realize that (dealing only with GLTFLoader objects in model-viewer). Will fix.